File: algorithms/Strings/Medium/ReorganizeString/VSDReorganizeString.py


import math
from collections import Counter

def issame(d):
    val=max(d.values())
    for i in d.values():
        if i!=val:
            return False
    return True 


def reorganizeString(string):
    d=dict()
    for i in string:
        if i in d.keys():
            d[i]+=1
        else:
            d[i]=1
    
    if max(d.values())>math.ceil(len(string)/2):
        return ""
    
    if issame(d):
        arr=list(sorted(string))
    else:
        arr=list(sorted(string,reverse=True,key=lambda x:d[x]))
    
    ans=[""]*len(string)
    
    j=0
    for i in range(0,len(arr),2):
        ans[i]=arr[j]
        j+=1
    for i in range(1,len(arr),2):
        ans[i]=arr[j]
        j+=1
    return ("").join(ans)


if __name__=="__main__":

    string="aab"
    print(reorganizeString(string))

File: algorithms/Strings/Medium/ReorganizeString/VSDReorganizeString.cpp
#include<bits/stdc++.h>

 std::string reorganizeString(std::string s) {
    
        std::unordered_map<char, int> mp;
        int mx=0;
        char ch;
        for(char c:s)
        {
            mp[c]++;
            if(mx<mp[c])
            {
                mx=mp[c];
                ch=c;
            }
        }
        int n=s.size();
        std::vector<char> v(n);
        int i=0;
        while(mx--)
        {
            if(i>=n)
            return "";
            v[i]=ch;
            i+=2;
        }
        for(auto it:mp)
        {
            if(it.first!=ch)
            {
                int k=it.second;
                while(k--)
                {
                    if(i>=n)
                    i=1;
                    v[i]=it.first;
                    i+=2;
                }
            }
        }
        std::string ans;
        for(char c:v)
        ans+=c;
        return ans;
}

int main(){
    std::string s="aab";
    std::cout<<reorganizeString(s);
    return 0;
}
 
